modules/light-clients/08-wasm/types/store.go

```diff
@@ -123,3 +129,46 @@ func (s storeAdapter) Iterator(start, end []byte) wasmvmtypes.Iterator {
 func (s storeAdapter) ReverseIterator(start, end []byte) wasmvmtypes.Iterator {
 	return s.parent.ReverseIterator(start, end)
 }
+
+func getClientID(clientStore storetypes.KVStore) (string, error) {
+	upws, isUpdateProposalWrappedStore := clientStore.(updateProposalWrappedStore)
+	if isUpdateProposalWrappedStore {
+		// if the clientStore is a updateProposalWrappedStore, we retrieve the subjectStore
+		// because the contract call will be made on the client with the ID of the subjectStore
+		clientStore = upws.subjectStore
+	}
+
+	store, ok := clientStore.(storeprefix.Store)
+	if !ok {
+		return "", errorsmod.Wrapf(ErrRetrieveClientID, "clientStore is not a prefix store")
+	}
+
+	// using reflect to retrieve the private prefix field
+	r := reflect.ValueOf(&store).Elem()
+
+	f := r.FieldByName("prefix")
+	if !f.IsValid() {
+		return "", errorsmod.Wrapf(ErrRetrieveClientID, "prefix field not found")
+	}
+
+	prefix := f.Bytes()
+	prefixStr := string(prefix)
+
+	lastSlash := strings.LastIndex(prefixStr, "/")
+	if lastSlash == -1 {
+		return "", errorsmod.Wrapf(ErrRetrieveClientID, "prefix does not contain a slash")
+	}
+	secondLastSlash := strings.LastIndex(prefixStr[:lastSlash], "/")
+	if secondLastSlash == -1 {
+		return "", errorsmod.Wrapf(ErrRetrieveClientID, "prefix does not contain a second slash")
+	}
+
+	clientID := prefixStr[secondLastSlash+1 : lastSlash]
+
+	isClientID := strings.HasPrefix(clientID, exported.Wasm)
+	if !isClientID {
+		return "", errorsmod.Wrapf(ErrRetrieveClientID, "prefix does not contain a %s clientID", exported.Wasm)
+	}
```

Comment:
I think this code could be simplified with `strings.Split` or something similar? I believe @chatton had some ideas.

Comments should indicate the expected format of the prefix

Reply:
[Carlos also had the idea of using regexps.](https://github.com/cosmos/ibc-go/pull/4939#discussion_r1372730697) Should we create an issue for this?

Reply:
I made some improvements